src/ophyd_async/epics/pvi/pvi.py

```diff
@@ -0,0 +1,280 @@
+    ),
+    frozenset({"r"}): lambda dtype, read_pv: epics_signal_r(dtype, "pva://" + read_pv),
+    frozenset({"w"}): lambda dtype, write_pv: epics_signal_w(
+        dtype, "pva://" + write_pv
+    ),
+    frozenset({"x"}): lambda _, write_pv: epics_signal_x("pva://" + write_pv),
+}
+
+
+def _parse_type(
+    is_pvi_table: bool,
+    number_suffix: Optional[int],
+    common_device_type: Optional[Type[Device]],
+):
+    if common_device_type:
+        # pre-defined type
+        device_type = _strip_union(common_device_type)
+        is_device_vector, device_type = _strip_device_vector(device_type)
+        import inspect
+
+        if ((origin := get_origin(device_type)) and issubclass(origin, Signal)) or (
+            inspect.isclass(device_type) and issubclass(device_type, Signal)
+        ):
+            # if device_type is of the form `Signal` or `Signal[type]`
+            is_signal = True
+            signal_dtype = get_args(device_type)[0]
+        else:
+            is_signal = False
+            signal_dtype = None
+
+    elif is_pvi_table:
+        # is a block, we can make it a DeviceVector if it ends ina number
+        is_device_vector = number_suffix is not None
+        is_signal = False
+        signal_dtype = None
+        device_type = Device
+    else:
+        # is a signal, signals aren't stored in DeviceVectors unless
+        # they're defined as such in the common_device_type
+        is_device_vector = False
+        is_signal = True
+        signal_dtype = None
+        device_type = Signal
+
+    return is_device_vector, is_signal, signal_dtype, device_type
+
+
+def _sim_common_blocks(device: Device, stripped_type: Optional[Type] = None):
+    device_t = stripped_type or type(device)
+    for sub_name, sub_device_t in get_type_hints(device_t).items():
+        if sub_name in ("_name", "parent"):
+            continue
+
+        # we'll take the first type in the union which isn't NoneType
+        sub_device_t = _strip_union(sub_device_t)
+        is_device_vector, sub_device_t = _strip_device_vector(sub_device_t)
+        is_signal = (
+            (origin := get_origin(sub_device_t)) and issubclass(origin, Signal)
+        ) or (issubclass(sub_device_t, Signal))
+
+        if is_device_vector and is_signal:
+            signal_type = args[0] if (args := get_args(sub_device_t)) else None
+            sub_device_1 = sub_device_t(SimSignalBackend(signal_type, sub_name))
+            sub_device_2 = sub_device_t(SimSignalBackend(signal_type, sub_name))
+            sub_device = DeviceVector(
+                {
+                    1: sub_device_1,
+                    2: sub_device_2,
+                }
+            )
+        elif is_device_vector and not is_signal:
+            sub_device = DeviceVector(
+                {
+                    1: sub_device_t(name=f"{device.name}-{sub_name}-1"),
+                    2: sub_device_t(name=f"{device.name}-{sub_name}-2"),
```

Comment:
Do the sub_devices need their prefixes passed too?

Reply:
Hmmm this is a good question. The pvs are given directly by the panda so I don't know if we actually need to have `_prefix` for anything :thinking:

@coretl should we be `lstrip`ing the given device PVs for their prefix at runtime?

Reply:
All PVs in PVI are always fully qualified, you only need the prefix to get to the root PVI structure, all child PVs do not need a prefix adding